cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java

```diff
@@ -670,6 +738,221 @@ public void testSearchMetrics() throws Exception {
+    ImmutableMap<String, ImmutableList<ImmutableList<QueryResult>>> expected =
+      ImmutableMap.of("testBatch", ImmutableList.<ImmutableList<QueryResult>>of(
+        ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(), "system.reads", 1, 3)),
+        ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(), "system.reads", 1, 1))));
+
+    q1 = getSingleQueryJson(ImmutableMap.of("namespace", "yourspace", "app", "WCount1", "flow", "WCounter",
+                                            "flowlet", "*"),
+                            ImmutableList.of("system.reads"), ImmutableList.<String>of(),
+                            ImmutableMap.of("aggregate", "true"));
+
+    q2 = getSingleQueryJson(ImmutableMap.of("namespace", "myspace", "app", "WordCount1", "flow", "WordCounter",
+                                            "flowlet", "splitter"),
+                            ImmutableList.of("system.reads", "system.writes"), ImmutableList.<String>of(),
+                            ImmutableMap.of("aggregate", "true"));
+    queries = new JsonArray();
+    queries.add(q1);
+    queries.add(q2);
+    batchTest(taggedQuery, expected);
+    taggedQuery = new JsonObject();
+    taggedQuery.add("testBatch2", queries);
+    expected = ImmutableMap.of("testBatch2", ImmutableList.<ImmutableList<QueryResult>>of(
+      ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(), "system.reads", 1, 4)),
+      ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(), "system.reads", 1, 2),
+                                    new QueryResult(ImmutableMap.<String, String>of(), "system.writes", 1, 2))
+    ));
+    batchTest(taggedQuery, expected);
+    taggedQuery = new JsonObject();
+    queries = new JsonArray();
+    queries.add(q1);
+    taggedQuery.add("testBatch1", queries);
+    queries = new JsonArray();
+    queries.add(q2);
+    taggedQuery.add("testBatch2", queries);
+
+    expected = ImmutableMap.of("testBatch1", ImmutableList.<ImmutableList<QueryResult>>of(
+                                 ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(),
+                                                                               "system.reads", 1, 4))),
+                               "testBatch2", ImmutableList.<ImmutableList<QueryResult>>of(
+        ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(), "system.reads", 1, 2),
+                                      new QueryResult(ImmutableMap.<String, String>of(), "system.writes", 1, 2)))
+    );
+    batchTest(taggedQuery, expected);
+
+    // test invalid request - query without any query Params and body content
+    HttpResponse response = doPost("/v3/metrics/query", null);
+    Assert.assertEquals(400, response.getStatusLine().getStatusCode());
+  }
+
+  JsonObject getSingleQueryJson(Map<String, String> tags, List<String> metrics,
```

Comment:
yeah, constructing json was a bad idea ;) moved QueryRequest to proto and using that now. Thanks